---
title: "Introduction to single-cell RNA-seq analysis"
subtitle: 'Normalisation Practical'
output:
  html_document:
    toc: yes
    toc_float: true
    toc_depth: 2
    number_sections: true
    css: ../css/boxes.css
    includes:
      in_header: ../css/navbar.html
---

```{r setup_wd}
knitr::opts_knit$set(root.dir = here::here("course_files"))
```

# Normalisation Practical

In the demonstration we ran the normalisation using just 500 cells per sample.
For this practical you will carry out normalisation, but this time using the 
all of the cells in one sample - ETV6-RUNX_1. Some of the commands will take a
little longer to run this time.

```{r setup, echo=FALSE, message=FALSE, warning=FALSE, purl=FALSE}
library(knitr)
library(scater)
library(scran)
library(sctransform)
library(tidyverse)
library(BiocParallel)

bpp <- MulticoreParam(7)
opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE, cache=TRUE)
set.seed(123)
```

```{r load_packages, eval=FALSE}
library(scater)
library(scran)
library(tidyverse)
library(BiocParallel)

bpp <- MulticoreParam(7)
```

## Prepare the data object

First we will load the full filtered data set. This data was previously filtered
to remove poor quality cells.

```{r load_data}
sce <- readRDS("R_objects/Caron_filtered.full.rds")
```

Now we need to extract just the cells for ETV6-RUNX_1.

```{r subset_data}
etvr1 <- which(sce$SampleName=="ETV6-RUNX1_1")
sce <- sce[, etvr1]
```

:::exercise 

Now that we have reduced the number of cells in the data set, it may be that
there are genes in the object that have not been detected in any of the 
remaining cells. Filter the object to remove any genes that have not been 
detected.

<details><summary>Hint</summary>

Any gene that has not been detected will have total UMI count across all cells
of zero. You can use `rowSums` on the `counts` matrix of the single cell object
to determine the total counts for each each gene. Then keep only rows with total
counts greater than 0.

</details>

<details><summary>Answer</summary>

```{r exercise_1_answer, purl=FALSE}
detected_genes <- rowSums(counts(sce)) > 0
sce <- sce[detected_genes,]
```

</details>

Q1. How many cells and genes are there in the ETV6_RUNX1_1 data set?

<details><summary>Answer</summary>

```{r exercise_1_Q1_answer_B, purl=FALSE}
dim(sce)
```

or

```{r exercise_1_Q1_answer_C, purl=FALSE}
sce
```


There are `r ncol(sce)` cells remaining and `r nrow(sce)` genes detected.

</details>

Q2. In how many cells has the gene ORC1 been detected with at least 1 UMI?

<details><summary>Hint</summary>

You will need to search the `rowData` of the single cell experiment 
object to determine which row of the `counts` matrix contains the data for the
gene ORC1. 

You can either use the row name (in this case the Ensembl ID) or the row number
to retrieve the counts for ORC1 from the counts matrix. You can then determine
how many of these are greater than 0.

</details>

<details><summary>Answer</summary>

```{r exercise_1_Q2, purl=FALSE}
orc1_row <- which(rowData(sce)$Symbol == "ORC1")
sum(counts(sce)[orc1_row, ] > 0)
```

</details>

:::

## Normalisation by deconvolution

:::exercise 

Now normalise the data set using the deconvolution method. You will need to

1. cluster the cells - remember to set a seed so that your results are
reproducible
2. compute size factors using the deconvolution method  
3. log normalize the counts by applying the size factors  
4. check that the single cell experiment object contains a new "logcounts" assay

<details><summary>Hint</summary>

These are the commands you will need to use for each step:

1. cluster the cells: `quickCluster`
2. compute size factors: `computePooledFactors`
3. log normalize the counts: `logNormCounts`
4. check assay: `assayNames`

</details>

<details><summary>Answer</summary>

### Cluster cells

```{r exercise_2_cluster_cells_answer, purl=FALSE}
set.seed(100) 
clust <- quickCluster(sce, BPPARAM = bpp)
```

### Compute size factors

```{r  exercise_2_compute_size_factors_answer, purl=FALSE}
sce <- computePooledFactors(sce,
			 clusters = clust,
			 min.mean = 0.1,
			 BPPARAM = bpp)
```

### Apply size factors

```{r  exercise_2_log_normalise_answer, purl=FALSE}
sce <- logNormCounts(sce)
```

### Check assays

```{r exercise_2_check_assays_answer, purl=FALSE}
assayNames(sce)
```

</details>

:::

## Normalisation with variance stabilising transformation

As in the demonstration, we will extract the counts matrix from our SCE object
and run the variance stabilising transformation (VST) algorithm, using the
`sctranform` package's `vst` command, directly on the matrix.

```{r extract_counts_matrix}
counts <- counts(sce)
```

### Estimation and transformation

:::exercise 

Now use the `vst` function on the counts matrix to estimate model parameters and
perform the variance stabilizing transformation. Call the newly created object
`vst_out`.

<details><summary>Hint</summary>

You will need to provide 4 arguments:

* `umi` - The matrix of UMI counts with genes as rows and cells as columns  
* `latent_var` - The independent variables to regress out as a character vector
- in this case this should be 'log_umi'
* `return_gene_attr` - set to 'TRUE'
* `return_cell_attr` - set to 'TRUE' 

</details>

<details><summary>Answer</summary>

```{r exercise_3_answer, eval=FALSE, purl=FALSE}
set.seed(44)
vst_out <- vst(umi = counts,
               latent_var = 'log_umi',
               return_gene_attr = TRUE,
               return_cell_attr = TRUE
  )
```


```{r exercise_3_answer_run, echo=FALSE, purl=FALSE}
set.seed(44)
vst_out <- vst(umi = counts,
               latent_var = 'log_umi',
               return_gene_attr = TRUE,
               return_cell_attr = TRUE,
               verbosity = 0
  )
```

</details>

:::

Check that the model used is the one we want (regressing out the effect of 
library size).

```{r check_vst_model}
print(vst_out$model_str)
```

:::exercise

Use the `plot_model` function to inspect the genes 'RPL10' and 'FTL' to see if
the fitted relationship between cell total UMI and gene expression, and to 
check the residuals.

<details><summary>Hint</summary>

First you will need to retrieve the Ensembl IDs for each of the genes from the
`rowData` of the original sce object.

The `plot_model` function requires four arguments:

+ `x` - the output of `vst`
+ `umi` - the counts matrix that `vst` was run on
+ `goi` - the IDs of the **g**enes **o**f **i**nterest - these should match the
rownames of the original counts matrix. In our case these should be Ensembl IDS
+ `plot_residual` - set this to `TRUE`

</details>

<details><summary>Answer</summary>

```{r exercise_4_answer, purl=FALSE}
ensId <- rowData(sce) %>%
	as.data.frame %>%
	filter(Symbol %in% c('RPL10', 'FTL')) %>%
  pull("ID")

plot_model(x = vst_out,
           umi = counts,
           goi = ensId,
           plot_residual = TRUE)
```

</details>

:::

### Overall properties of transformed data

The distribution of residual mean should be centered around 0.

```{r residual_mean_histogram}
ggplot(vst_out$gene_attr, aes(residual_mean)) +
  geom_histogram(binwidth=0.01)
```

The distribution of residual variance should be centered around 1.

```{r residual_variance_histogram}
ggplot(vst_out$gene_attr, aes(residual_variance)) +
  geom_histogram(binwidth=0.1) +
  geom_vline(xintercept=1, color='red') +
  xlim(0, 10)
```

Finally, check that the relationship between variance and mean has been removed.

```{r residual_variance_v_mean_plot}
ggplot(vst_out$gene_attr,
       aes(log10(gmean), residual_variance)) +
       geom_point(alpha=0.3, shape=16)
```
